exec/java-exec/src/main/java/org/apache/drill/exec/ops/QueryContext.java

```diff
@@ -144,6 +145,25 @@ public SchemaPlus getNewDefaultSchema() {
     return defaultSchema;
   }
 
+  public SchemaPlus getPartialDefaultSchema() {
+
+    final String sessionSchemaPath = session.getDefaultSchemaPath();
+    final List<String> schemaPathAsList = Lists.newArrayList(sessionSchemaPath.split("\\."));
+
+    final SchemaPlus rootSchema = schemaTreeProvider.createPartialRootSchema(getQueryUserName(),
+        this, schemaPathAsList.get(0));
+
+    final SchemaPlus defaultSchema = session.getDefaultSchema(rootSchema);
+
+    if (defaultSchema == null) {
+      return rootSchema;
+    }
+    return defaultSchema;
+  }
+
+  public void addNewRelevantSchema(Set<String> storages, SchemaPlus toExpandSchema) {
+    schemaTreeProvider.addPartialRootSchema(getQueryUserName(), this, storages, toExpandSchema);
```

Comment:
Per a comment later on, find a good home for computing the effective user name (with impersonation enabled). That calculation should probably be done on the user session if the impersonation is for the life of the session, or on the query context if the impersonation is only for the single query. (Note that if it is only for a query, temp tables probably won't work the way we expect.)

But, doing the impersonation check in the schema tree provider is pushing that behavior too far down into the code.